packages/protocol-kit/src/SafeProvider.ts

```diff
@@ -32,49 +22,114 @@ import {
   SafeProviderConfig,
   Eip1193Provider,
   HttpTransport,
-  SocketTransport
+  SocketTransport,
+  ExternalSigner
 } from '@safe-global/protocol-kit/types'
+import { asAddress, asHash, asHex, getChainById } from './utils/types'
+import {
+  createPublicClient,
+  createWalletClient,
+  PublicClient,
+  custom,
+  http,
+  getAddress,
+  isAddress,
+  Transaction,
+  decodeAbiParameters,
+  encodeAbiParameters,
+  parseAbiParameters,
+  toBytes,
+  BlockTag,
+  Chain
+} from 'viem'
+import { privateKeyToAccount } from 'viem/accounts'
+import {
+  toEstimateGasParameters,
+  toCallGasParameters,
+  sameString
+} from '@safe-global/protocol-kit/utils'
+
+function asBlockId(blockId: number | string | undefined) {
+  return typeof blockId === 'number' ? blockNumber(blockId) : blockTag(blockId)
+}
+
+function blockNumber(blockNumber: any) {
+  return { blockNumber: blockNumber.toNumber() }
+}
+
+function blockTag(blockTag: any) {
+  return { blockTag: blockTag as BlockTag }
+}
 
 class SafeProvider {
-  #externalProvider: BrowserProvider | JsonRpcProvider
+  #chain?: Chain
+  #externalProvider: PublicClient
   signer?: string
   provider: Eip1193Provider | HttpTransport | SocketTransport
 
   constructor({ provider, signer }: SafeProviderConfig) {
     if (typeof provider === 'string') {
-      this.#externalProvider = new JsonRpcProvider(provider)
+      this.#externalProvider = createPublicClient({
```

Comment:
I think perhaps this logic could be directly used in the getExternalProvider. But more than that, I think we can rename getExternalProvider and getExternalSigner to getPublicClient and getWalletClient as it is more viem friendly for the developers to understand what it returns WDYT?

Reply:
I think it would make sense, but it would also introduce additional breaking changes which we tried to keep at a minimum on this PR. So I think it would be better to do this in separate PR.